aws/callback.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as aws from "@pulumi/aws";
+import * as cloud from "@pulumi/cloud";
+import * as pulumi from "@pulumi/pulumi";
+import { functionIncludePackages, functionIncludePaths, functionMemorySize } from "./config";
+import { getComputeIAMRolePolicies, getOrCreateNetwork, runLambdaInVPC } from "./shared";
+
+/**
+ * AWS-specific data to create an AWS lambda out of a callback function.  Can be passed to any
+ * functions in pulumi/cloud-aws that can take a cloud.Callback<T> argument.  To create the same
+ * default AwsCallbackData Pulumi creates when given a simple JavaScript function, use
+ * [createCallbackData].
+ */
+export interface AwsCallbackData<T extends Function> extends cloud.CallbackData<T>, aws.lambda.CallbackFunctionArgs<any, any> {
+    /**
+     * Not used.  Provide [function] instead.
+     */
+    callback?: never;
+    /**
+     * Not used.  Provide [function] instead.
+     */
+    callbackFactory?: never;
+}
+
+/**
+ * Type for parameters that will be converted into serverless function.  Either a simple JavaScript
+ * function, or an object with information necessary to create an AWS Lambda can be used.
+ */
+export type AwsCallback<T extends Function> = T | AwsCallbackData<T>;
+
+export function createCallbackData<T extends Function>(func: T): AwsCallbackData<T> {
```

Comment:
Will document this, but i think it's important ot be publicly exported.  It enables users to create a CallBackData instance htat matches how we do things, which they can then override individual properties on.  Without this, they'd have to duplicate this logic (somehow) if they wanted things like running in a VPC.